metrics-publishers/cloudwatch-metrics-publisher/src/main/java/software/amazon/awssdk/metrics/publishers/cloudwatch/CloudWatchMetricsPublisher.java

```diff
@@ -0,0 +1,258 @@
+
+    private final AtomicBoolean publishStarted = new AtomicBoolean(false);
+    private final Duration publishFrequency;
+
+    private CloudWatchMetricsPublisher(Builder builder) {
+        this.client = resolveClient(builder.client, builder.awsCredentialsProvider, builder.region);
+        this.publishFrequency = Validate.notNull(builder.publishFrequency, "Publish frequency cannot be null.");
+        this.namespace = Validate.notEmpty(builder.namespace, "Namespace cannot be null or empty.");
+        this.metricQueueSize = Validate.isPositive(builder.metricQueueSize, "Metric queue size should be positive.");
+        this.queue = new LinkedBlockingQueue<>(this.metricQueueSize);
+
+        this.producer = MetricProducer.builder().queue(queue).build();
+        this.producerExecutorService = Executors.newSingleThreadExecutor(new ThreadFactoryBuilder()
+                                                                             .threadNamePrefix("sdk-metrics-cw-producer")
+                                                                             .build());
+
+        this.consumer = MetricConsumer.builder()
+                                      .cloudWatchClient(client)
+                                      .queue(queue)
+                                      .namespace(namespace)
+                                      .build();
+        this.consumerExecutorService = Executors.newSingleThreadScheduledExecutor(new ThreadFactoryBuilder()
+                                                                                      .threadNamePrefix("sdk-metrics-cw-consumer")
+                                                                                      .build());
+    }
+
+    /**
+     * CloudWatch publisher converts the given metrics into {@link MetricDatum} instances and add them
+     * to a queue for publishing.
+     *
+     * @param metricsRegistry registry containing the collected metrics
+     */
+    @Override
+    public void registerMetrics(MetricRegistry metricsRegistry) {
+        try {
+            if (publishStarted.compareAndSet(false, true)) {
+                consumerExecutorService.scheduleAtFixedRate(this::publish,
+                                                            0L,
+                                                            publishFrequency.toMillis(),
+                                                            TimeUnit.MILLISECONDS);
+            }
+
+            producerExecutorService.execute(() -> producer.addMetrics(metricsRegistry));
+        } catch (Throwable throwable) {
+            log.warn(() -> "An error occurred when registering metrics in the publisher", throwable);
+        }
+    }
+
+    @Override
+    public CompletableFuture<Void> publish() {
+        List<CompletableFuture<PutMetricDataResponse>> futures = new ArrayList<>();
+        int count = 0;
+
+        try {
+            while (queue.peek() != null && ++count <= MAX_SERVICE_CALLS_PER_PUBLISH) {
```

Comment:
What do we think about moving the queue in the `MetricConsumer` class? This class doesn't need to have knowledge of the queue at all and two classes having access to the same queue might be problematic in the future.

Reply:
As in move this logic to the MetricConsumer class? Not sure I follow otherwise.

Reply:
Yeah, I meant moving the creation of the queue and this logic into `MetricConsumer` class.

Reply:
The MetricProducer also uses the queue.

Reply:
Oh I missed that part. Do we need to check `queue.peek` here? can we move this logic to consumer?